doc/dev_guide/c_style_guide_new.rst

```diff
@@ -0,0 +1,1046 @@
+Avoid using GW-BASIC names like ``err1:`` and ``err2:``, as you would have to
+renumber them if you ever add or remove exit paths, and they make correctness
+difficult to verify anyway.
+
+The rationale for using gotos is:
+
+- unconditional statements are easier to understand and follow
+- nesting is reduced
+- errors by not updating individual exit points when making
+  modifications are prevented
+- saves the compiler work to optimize redundant code away ;)
+
+.. code-block:: c
+
+  int
+  fun(int a)
+  {
+    int result = 0;
+    char *buffer;
+
+    buffer = kmalloc(SIZE, GFP_KERNEL);
+    if (!buffer)
+      return -ENOMEM;
+
+    if (condition1) {
+      while (loop1) {
+        ...
+      }
+      result = 1;
+      goto out_free_buffer;
+    }
+    ...
+  out_free_buffer:
+    kfree(buffer);
+    return result;
+  }
+
+A common type of bug to be aware of is ``one err bugs`` which look like this:
+
+.. code-block:: c
+
+  err:
+    kfree(foo->bar);
+    kfree(foo);
+    return ret;
+
+The bug in this code is that on some exit paths ``foo`` is NULL.  Normally the
+fix for this is to split it up into two error labels ``err_free_bar:`` and
+``err_free_foo:``:
+
+.. code-block:: c
+
+   err_free_bar:
+    kfree(foo->bar);
+   err_free_foo:
+    kfree(foo);
+    return ret;
+
+Ideally you should simulate errors to test all exit paths.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Chapter 8: Commenting
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Comments are good, but there is also a danger of over-commenting.  NEVER
+try to explain HOW your code works in a comment: it's much better to
+write the code so that the **working** is obvious, and it's a waste of
+time to explain badly written code.
+
+Generally, you want your comments to tell WHAT your code does, not HOW.
+Also, try to avoid putting comments inside a function body: if the
+function is so complex that you need to separately comment parts of it,
+you should probably go back to chapter 6 for a while. You can make
+small comments to note or warn about something particularly clever (or
+ugly), but try to avoid excess. Instead, put the comments at the head
+of the function, telling people what it does, and possibly WHY it does
+it.
+
+.. admonition:: Tarantool Style
+    :class: FACT
+
+    When commenting the Tarantool C API functions, please use Doxygen comment format,
+    Javadoc flavor, i.e. `@tag` rather than `\\tag`.
+    The main tags in use are ``@param``, ``@retval``, ``@return``, ``@see``,
+    ``@note`` and ``@todo``.
+
+    Every function, except perhaps a very short and obvious one, should have a
+    comment. A sample function comment may look like below:
+
+    .. code-block:: c
+
+        /** Write all data to a descriptor.
+         *
+         * This function is equivalent to 'write', except it would ensure
+         * that all data is written to the file unless a non-ignorable
+         * error occurs.
+         *
+         * @retval 0  Success
+         *
```

Comment:
Don't need to put empty lines between doxygen commands. At least if they are short.